#399 #416 Initial fix for header/footer showing on page without body rows

danfickle opened the pull request:
I also plan to add some commits to this PR with general table cleanup of code.

@VictorAtPL, if you get time, you could perhaps test this branch with your real world tables.

danfickle committed:
#399 #416 Initial fix for header/footer showing on a page with no body rows.

With test proof.

danfickle committed:
Compiler fixes for newtable package. [ci skip]

+ Remove unneeded casts.
+ Add override annotations.
+ Change anonymous class creation to lambda (manual).

danfickle committed:
#202 #399 Failing test for table with too high first body row. [ci skip]

Plus orphaned thead.

Also compiler fixes for ContentLimit/ContentLimitContainer.

Test based on the one provided by @jesselong. Thanks.

danfickle commented:
The test I just committed for #202 still produces this:

![issue-202-malformed-table](https://user-images.githubusercontent.com/1415728/69475116-4f918a00-0e1d-11ea-9393-f6c8a7db48da.png)

so clearly there is still more related problems in the table code.

VictorAtPL commented:
Right now it works differently. It puts part of text on the previous page, and part of it on the next, even that I have:
```
div.parts table.part tr, div.parts table.part tr td {
    page-break-inside: avoid;
}
```
in the styles.
![openhtmltopdf_result_after_authors_bugfix](https://user-images.githubusercontent.com/1886347/69480131-db4aec80-0e04-11ea-8b37-4a9bfc923aaf.png)

danfickle committed:
Merge branch 'open-dev-v1' into fix_399_empty_table_header_footer

danfickle committed:
#202 - Correct rendering for example.

Need more testing before we can merge this branch.
Specifically:
Cells with margin/border/padding.
Cells which go over two or more pages.

danfickle committed:
#202 Two more paged table tests with partial revert to make them pass.

Partial revert of dce1cd54 to get paged table with cells with large padding/border rendering correctly.

Unfortunately, this means that the #202 test with explicitly allowing page breaks in thead now fails again.

Fortunately, very few users want a page break in their thead so this is low priority to fix.

leedsalex commented:
@danfickle Hi, any update when this will be ready to be released? I'm hopeful it will fix the issues i'm experiencing.

danfickle commented:
@VictorAtPL, it is not clear from your comment if this is a problem introduced in this pr or was already existing? If it was already existing I propose to merge this pr and do a release, then we can work more on table rendering in a different pr.

@leedsalex Very soon if this pr has not introduced any additional problems as above.

VictorAtPL commented:
@danfickle 
I will test it tomorrow and answer your question.

VictorAtPL commented:
@danfickle
I am not sure. 

v. 1.0.1:
![image](https://user-images.githubusercontent.com/1886347/73135532-89605600-4043-11ea-82ee-5ae7ddc092f1.png)
![image](https://user-images.githubusercontent.com/1886347/73135534-92512780-4043-11ea-9e83-f72c06b022db.png)

fix_399_empty_table_header_footer
![image](https://user-images.githubusercontent.com/1886347/73135552-b7459a80-4043-11ea-874f-8b617b451bc7.png)
![image](https://user-images.githubusercontent.com/1886347/73135558-c4fb2000-4043-11ea-8167-bb2e14de4aa7.png)

https://github.com/VictorAtPL/openhtmltopdf/tree/fix_thead_tfoot_without_content_bug (my dirty fix)
![image](https://user-images.githubusercontent.com/1886347/73135571-f2e06480-4043-11ea-84c4-3e0573bf28c7.png)
![image](https://user-images.githubusercontent.com/1886347/73135573-faa00900-4043-11ea-8c05-300c21a01895.png)

As you can see, there are still problems with whole header and footer being not rendered ("rectangle" near "Grafika" and arrow below "C"). Also, on your branch, "dsadsa" of second C is on the previous page, and image of the second C doesn't have margin.

danfickle commented:
Hi @VictorAtPL,
I'm back finally. I had a play around but wasn't able to replicate. Are you able to share how the left blue arrow bar is achieved? At a guess it involves rowspan, transforms and a CSS arrow? Even better would be a sample.
